ql/src/java/org/apache/hadoop/hive/ql/optimizer/SharedWorkOptimizer.java

```diff
@@ -486,6 +510,75 @@ private static boolean sharedWorkOptimization(ParseContext pctx, SharedWorkOptim
     return mergedExecuted;
   }
 
+  private static void replaceSemijoinExpressions(TableScanOperator tsOp, List<ExprNodeDesc> semijoinExprNodes) {
```

Comment:
We are merging the discardable TS into the retainable TS. We remove the SJ expressions from the discardable TS since we are about to remove it from the operator tree (we preserve the rest of filter expressions without SJ expressions from this branch + the expressions with SJ coming from other branch so we can add them to the Filter operator on top of the TS).

Reply:
So the SJ branch from retainable TS is pushed to discardable TS  so that it is later pushed to discardable TS's children filter op? Why do we want to do that? Shouldn't retainable SJ be retained with retainable FILTER op only?

Reply:
Yes, that is correct. But it is not the branch, it is only the expressions.

The reason is that we have already verified that the SJ expressions hitting both TS operators are the same. At the current step we are already merging. Thus, what we want now is that the SJ expression from the retainable branch is on top of the discardable branch too. Since we already had a method to push the filter expressions on top of the discardable TS (`pushFilterToTopOfTableScan`), what I have done is that we remove the old SJ expressions from the discardable TS (and follow-up Filters if present) and we add the SJ expressions from the retainable TS, hence automatically they will be pushed on top of the discardable TS. Then we can just remove the discardable TS operator and connect its output operators with the retainable TS operator.

Reply:
This makes sense. Thanks for the explanation. Would you mind adding a comment to `replaceSemijoinExpressions`?